### PR TITLE
Added Ideal and Robust Soliton Distributions

### LINKED ISSUE #20100: Soliton Distributions
I would like to add the Ideal and Robust Soliton Distributions to the stats module. More on this here - 
[Wikipedia](https://en.wikipedia.org/wiki/Soliton_distribution)
[Luby's Research Paper](http://pages.cs.wisc.edu/~suman/courses/740/papers/luby02lt.pdf)

### PR DESCRIPTION
**Pull request by naveensaigit**
#### References to other Issues or PRs
Fixes #20100 

#### Brief description of what is fixed or changed

The soliton distributions are used in the theory of erasure correcting codes, which use information redundancy to compensate for transmission errors manifesting as missing data. I have implemented both Ideal and Robust Soliton Distributions in the FRV section of stats module

#### Release Notes

* stats
   * Added 2 new distributions in frv_types.

### CODE REVIEW

**Comment by naveensaigit** [comment]
@sylee957 @asmeurer Could someone review this PR?

File: sympy/core/tests/test_args.py [hunk 1]
```diff
@@ -1271,6 +1271,13 @@ def test_sympy__stats__frv_types__RademacherDistribution():
     from sympy.stats.frv_types import RademacherDistribution
     assert _test_args(RademacherDistribution())
 
+def test_sympy__stats__frv_types__IdealSolitonDistribution():
+    from sympy.stats.frv_types import IdealSolitonDistribution
+    assert _test_args(IdealSolitonDistribution(10))
+
+def test_sympy__stats__frv_types__RobustSolitonDistribution():
+    from sympy.stats.frv_types import RobustSolitonDistribution
+    assert _test_args(RobustSolitonDistribution(1000,0.5,0.1))
```
Comment (Smit-create) [inline sympy/core/tests/test_args.py:1280]:
```suggestion
    assert _test_args(RobustSolitonDistribution(1000, 0.5, 0.1))
```

File: sympy/stats/frv_types.py [hunk 2]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
 
     """
     return rv(name, RademacherDistribution)
+
+class IdealSolitonDistribution(SingleFiniteDistribution):
+    _argnames = ('k',)
+
+    @staticmethod
+    def check(k):
+         _value_check((k.is_integer, k.is_positive),
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:659]:
```suggestion
         _value_check(k.is_integer and k.is_positive,
```

File: sympy/stats/frv_types.py [hunk 3]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
...
+    def low(self):
+        return S.One
+
+    @property
+    def high(self):
+        return self.k
+
+    @property
+    def set(self):
+        return set(list(range(1,self.k+1)))
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:672]: This should be changed to sympy's `Range` as python's `range` will not work for symbolic k

File: sympy/stats/frv_types.py [hunk 4]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
...
+
+    @property
+    def is_symbolic(self):
+        return not self.k.is_number
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:676]: This isn't necessary to add. You can simply check using `k.is_Symbol`

File: sympy/stats/frv_types.py [hunk 5]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
...
+
+    @property
+    @cacheit
+    def dict(self):
+        if self.is_symbolic:
+            return Density(self)
+        d={1:Rational(1,self.k)}
+        d.update(dict((i,Rational(1,i*(i-1))) for i in range(2,self.k+1)))
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:684]:
```suggestion
        d={1: Rational(1, self.k)}
        d.update(dict((i, Rational(1, i*(i - 1))) for i in range(2, self.k + 1)))
```

File: sympy/stats/frv_types.py [hunk 6]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
...
+        return d
+
+    def pmf(self,x):
+        x=sympify(x)
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:688]:
```suggestion
    def pmf(self, x):
        x = sympify(x)
```

File: sympy/stats/frv_types.py [hunk 7]
```diff
@@ -646,3 +650,204 @@ def Rademacher(name):
...
+        if not (x.is_number or x.is_Symbol or is_random(x)):
+            raise ValueError("'x' expected as an argument of type 'number' or 'Symbol' or , "
+                        "'RandomSymbol' not %s" % (type(x)))
+        cond1 = Eq(x,1) & x.is_integer
+        cond2 = Ge(x,1) & Le(x,self.k) & x.is_integer
+        return Piecewise((1/self.k,cond1),(1/(x*(x-1)),cond2),(S.Zero,True))
```
Comment (Smit-create) [inline sympy/stats/frv_types.py:694]: Add space after commas. See above suggestions

**Comment by naveensaigit** [comment]
@Smit-create I've made the requested changes. If you could review the RobustSoliton as well, that'd be great!

**Comment by Smit-create** [comment]
Can you please show the behavior of `FiniteRV` using the `pmf` of RobustSoliton and Ideal Distribution. If we are able to get the expected results using `FiniteRV` then in my opinion we should not add them and let FiniteRV handle it. Please show a session with FiniteRV and its results on `E`, `P`, `density`, `cdf` ,etc.

**Comment by naveensaigit** [comment]
> Can you please show the behavior of FiniteRV using the pmf of RobustSoliton and Ideal Distribution

You can simulate both the distributions for small values of `k`, but these distributions are used where the value of `k` is around 10000. So, using `FiniteRV` hardly makes sense. You can go through this link to understand the applications of these distributions - [Digital Fountain Codes](http://www.inference.org.uk/mackay/itprnn/ps/588.596.pdf) (See pg 5 of this research paper)

**Comment by Smit-create** [comment]
>  So, using `FiniteRV` hardly makes sense.

If that is the case then I think you should look into fixing `FiniteRV`, as I don't think there should be any performance difference between the `FiniteRV` or Direct Distribution. If `FiniteRV` is unable to produce the result for k=1e5 (for Ideal Soliton), can you show that this diff outperforms `FiniteRV` with k=1e5 ?

**Comment by naveensaigit** [comment]
> If that is the case then I think you should look into fixing `FiniteRV`, as I don't think there should be any performance difference between the `FiniteRV` or `Direct Distribution`.

I didn't mean to say that `FiniteRV` isn't efficient for the distribution. All I meant to say this that the user has to calculate the pmf for these distributions and thus, the user has to code these distributions from scratch

**Comment by naveensaigit** [comment]
@Smit-create Can you merge this or are there any other changes required?

**Comment by oscarbenjamin** [comment]
Does `FiniteRV` work if the size of the statespace is only symbolically known e.g.:
```julia
In [44]: n = Symbol('n', integer=True, positive=True)                                                                                          

In [45]: B = Binomial('B', n, S.Half) 
```

File: sympy/stats/frv_types.py [hunk 8]
```diff
@@ -646,3 +650,200 @@ def Rademacher(name):
...
+         _value_check(k.is_integer and k.is_positive,
+                    "'k' must be a positive integer.")
+
+    @property
+    def low(self):
+        return S.One
+
+    @property
+    def high(self):
+        return self.k
```
Comment (czgdp1807) [inline sympy/stats/frv_types.py:668]: Do we need this? I don't remember if there were such properties in other distribution classes.

**Comment by czgdp1807** [comment]
> Does `FiniteRV` work if the size of the statespace is only symbolically known e.g.:

No, they don't seem to in current `master`.

```python
In [1]: from sympy.stats import *

In [2]: n = Symbol('n', integer=True, positive=True)

In [3]: B = Binomial('B', n, S.Half)

In [4]: P(B > 2)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-4-aacaa9e50781> in <module>
----> 1 P(B > 2)

~/sympy_project/sympy/sympy/stats/rv.py in probability(condition, given_condition, numsamples, evaluate, **kwargs)
    812     from sympy.stats.symbolic_probability import Probability
    813     if evaluate:
--> 814         return Probability(condition, given_condition).doit(**kwargs)
    815     ### TODO: Remove the user warnings in the future releases
    816     message = ("Since version 1.7, using `evaluate=False` returns `Probability` "

~/sympy_project/sympy/sympy/stats/symbolic_probability.py in doit(self, **hints)
    109             return Probability(condition, given_condition)
    110 
--> 111         result = pspace(condition).probability(condition)
    112         if hasattr(result, 'doit') and for_rewrite:
    113             return result.doit()

~/sympy_project/sympy/sympy/stats/frv.py in probability(self, condition)
    542         if self._is_symbolic:
    543             #TODO: Implement the mechanism for handling queries for symbolic sized distributions.
--> 544             raise NotImplementedError("Currently, probability queries are not "
    545             "supported for random variables with symbolic sized distributions.")
    546         condition = rv_subs(condition)

NotImplementedError: Currently, probability queries are not supported for random variables with symbolic sized distributions.
```

File: sympy/stats/frv_types.py (hunk 8 above)
Comment (naveensaigit) [reply sympy/stats/frv_types.py:668]: Yes, discrete random variables have them defined, so I added them

**Comment by czgdp1807** [comment]
A bunch of lines [in this report](https://codecov.io/gh/sympy/sympy/pull/20102/diff) haven't been covered. @naveensaigit Could you please add some tests for them?

**Comment by naveensaigit** [comment]
It looks like I deleted the branch sometime back. I've cloned the branch using the SHA and made changes too, but when I push, a new branch is being created, but there is no effect here. Should I open a new PR, or is there a way to restore this?

**Comment by czgdp1807** [comment]
May be you can restore the deleted branch here on github(if you haven't permanently deleted from github) and then fetch it on your local machine, make the changes and re-push it(may be force push if normal push resists).

**Comment by naveensaigit** [comment]
I was unable to make changes appear here. However, the branch is restored and all the commit history is intact. Please see #20514. Closing this now

**Pull request closed without merging**